src/array/utf8/mutable.rs

```diff
@@ -223,7 +223,145 @@ impl<O: Offset, P: AsRef<str>> FromIterator<Option<P>> for MutableUtf8Array<O> {
+        assert_eq!(array.offsets().as_slice(), &[0, 2, 7, 7, 12]);
+        assert_eq!(
+            array.validity(),
+            &Some(Bitmap::from_u8_slice(&[0b00001011], 4))
+        );
+    }
+}
+
 impl<O: Offset> MutableUtf8Array<O> {
+    /// Extends the [`MutableUtf8Array`] from an iterator of values of trusted len.
+    /// This differs from `extended_trusted_len` which accepts iterator of optional values.
+    #[inline]
+    pub fn extend_trusted_len_values<I, P>(&mut self, iterator: I)
+    where
+        P: AsRef<str>,
+        I: TrustedLen<Item = P>,
+    {
+        unsafe { self.extend_trusted_len_values_unchecked(iterator) }
+    }
+
+    /// Extends the [`MutableUtf8Array`] from an iterator of values of trusted len.
+    /// This differs from `extended_trusted_len_unchecked` which accepts iterator of optional
+    /// values.
+    /// # Safety
+    /// The iterator must be trusted len.
+    #[inline]
+    pub unsafe fn extend_trusted_len_values_unchecked<I, P>(&mut self, iterator: I)
+    where
+        P: AsRef<str>,
+        I: Iterator<Item = P>,
+    {
+        let (_, upper) = iterator.size_hint();
+        let additional = upper.expect("extend_trusted_len_values requires an upper limit");
+
+        // TODO How to avoid creating vec, how to assign capacity without iterating
+        // the entire input iterator
+        let mut values = vec![];
+        let mut offsets = Vec::with_capacity(additional);
+
+        let mut length = self.last_offset();
+
+        for item in iterator {
+            let bytes = item.as_ref().as_bytes();
+
+            length += O::from_usize(bytes.len()).unwrap();
+
+            values.extend_from_slice(bytes);
+            offsets.push(length);
+        }
+
+        self.values
+            .extend_from_trusted_len_iter_unchecked(values.into_iter());
+        self.offsets
+            .extend_from_trusted_len_iter_unchecked(offsets.into_iter());
+
+        if let Some(validity) = self.validity.as_mut() {
+            validity.extend_constant(additional, true);
+        }
+    }
+
+    /// Extends the [`MutableUtf8Array`] from an iterator of trusted len.
+    #[inline]
+    pub fn extend_trusted_len<I, P>(&mut self, iterator: I)
+    where
+        P: AsRef<str>,
+        I: TrustedLen<Item = Option<P>>,
+    {
+        unsafe { self.extend_trusted_len_unchecked(iterator) }
+    }
+
+    /// Extends [`MutableUtf8Array`] from an iterator of trusted len.
+    /// #Safety
+    /// The iterator must be trusted len.
+    #[inline]
+    pub unsafe fn extend_trusted_len_unchecked<I, P>(&mut self, iterator: I)
+    where
+        P: AsRef<str>,
+        I: Iterator<Item = Option<P>>,
+    {
+        let (_, upper) = iterator.size_hint();
+        let additional = upper.expect("extend_trusted_len_unzip requires an upper limit");
+
+        // TODO How to reserve space for `values` and `validity` and take advantage
+        // of `push_unchecked`
+        self.offsets.reserve(additional);
+
```

Comment:
@jorgecarleitao Is it possible to deterministically assign reserves for `values` and `validity`?

`validity` is handled differently here than `MutablePrimitiveArray`, here we avoid allocating any space for it if all our inputs are valid, is this correct?